persist,table: refactor Tables and SourceDesc to make clear what the persisted_name does

aljoscha opened the pull request:
Before, the `persist_name` on SourceDesc was serving two purposes: 1) to
signal that a source actively persists, and 2) to mark a Table as a
persistent table that gets its data by creating a `persisted_source()`
with the given `persist_name`.

Now we clearly separate the two and move the `persist_name` into
`SourceConnector::Local`. It becomes a property of the local source
instead of the descriptor around it.

This will help with future refactoring that change the shape of the
persistence information that we ship with `SourceDesc`.

### Tips for reviewer

I left notes and a potential future `TODO` in the code.


### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [N/A] This PR adds a release note for any [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/user/content/release-notes.md#what-changes-require-a-release-note).

danhhz commented:
I was pretty blindly stabbing around when I added persist_name where it is, so I'm find moving it. By the same reasoning, someone other than me should probably be the one to give the stamp here :)

aljoscha commented:
> I was pretty blindly stabbing around when I added persist_name where it is, so I'm find moving it.

It is only like this now since I added the persistent Source... :see_no_evil: 

> By the same reasoning, someone other than me should probably be the one to give the stamp here.

I don't know if such a person exists, I'll ask around.

aljoscha commented:
Btw, my thinking on this is similar to how we now think of the two halves of `(IngestdSourceConnector, DataflowdSourceConnector)`. Where `SourceConnector::Local {persisted_name}` is the dual to the machinery on `coord` that writes to the persistent collection that this replays/listens on.

aljoscha commented:
Oh no, the shame, I'm sorry :see_no_evil: I forget that it was me who moved this: https://github.com/MaterializeInc/materialize/pull/8746/commits/d024415fc466302f3c4d7c03db1c39f2df8e5402. At least now I have a clearer opinion about why it should be there. Silver linings... :sweat_smile: 

So we can probably just merge?

danhhz reviewed (approved):
WFM!
